javascript/ql/src/Security/CWE-020/IncompleteHostnameRegExp.ql

```diff
@@ -0,0 +1,67 @@
+/**
+ * @name Incomplete regular expression for hostnames
+ * @description Matching a URL or hostname against a regular expression that contains an unescaped dot as part of the hostname might match more than expected.
+ * @kind problem
+ * @problem.severity warning
+ * @precision high
+ * @id js/incomplete-hostname-regexp
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-20
+ */
+
+import javascript
+
+module IncompleteHostnameRegExpTracking {
+
+  /**
+   * A taint tracking configuration for incomplete hostname regular expressions sources.
+   */
+  class Configuration extends TaintTracking::Configuration {
+    Configuration() { this = "IncompleteHostnameRegExpTracking" }
+
+    override
+    predicate isSource(DataFlow::Node source) {
+      isIncompleteHostNameRegExpPattern(source.asExpr().getStringValue(), _)
+    }
+
+    override
+    predicate isSink(DataFlow::Node sink) {
+      isInterpretedAsRegExp(sink)
+    }
+
+  }
+
+}
+
+/**
+ * Holds if `pattern` is a regular expression pattern for URLs with a host matched by `hostPart`,
+ * and `pattern` contains a subtle mistake that allows it to match unexpected hosts.
+ */
+bindingset[pattern]
+predicate isIncompleteHostNameRegExpPattern(string pattern, string hostPart) {
+  hostPart = pattern.regexpCapture(
+    "(?i).*" +
+    // an unescaped single `.`
+    "(?<!\\\\)[.]" +
+    // immediately followed by a sequence of subdomains, perhaps with some regex characters mixed in, followed by a known TLD
+    "([():|?a-z0-9-]+(\\\\)?[.](com|org|edu|gov|uk|net))" +
```

Comment:
IIRC we now have two copies of this "well-known TLD" (sub-)regexp. Time to pull it out into the library? Also, how did you decide which TLDs go into this list?

Reply:
Moved into: `Regexp::RegExpPatterns::commonTLD`.